crates/bevy_ecs/src/component.rs

```diff
@@ -247,20 +269,41 @@ impl Components {
     #[inline]
     pub fn init_component<T: Component>(&mut self, storages: &mut Storages) -> ComponentId {
         let type_id = TypeId::of::<T>();
-        let components = &mut self.components;
-        let index = self.indices.entry(type_id).or_insert_with(|| {
-            let index = components.len();
-            let descriptor = ComponentDescriptor::new::<T>();
-            let info = ComponentInfo::new(ComponentId(index), descriptor);
-            if T::Storage::STORAGE_TYPE == StorageType::SparseSet {
-                storages.sparse_sets.get_or_insert(&info);
-            }
-            components.push(info);
-            index
+
+        let Components {
+            indices,
+            components,
+            ..
+        } = self;
+        let index = indices.entry(type_id).or_insert_with(|| {
+            Components::init_component_inner(components, storages, ComponentDescriptor::new::<T>())
         });
         ComponentId(*index)
     }
 
+    pub fn init_component_with_descriptor(
+        &mut self,
+        storages: &mut Storages,
+        descriptor: ComponentDescriptor,
+    ) -> ComponentId {
+        let index = Components::init_component_inner(&mut self.components, storages, descriptor);
+        ComponentId(index)
+    }
+
+    fn init_component_inner(
```

Comment:
Should this be `#[inline]` like `init_component`?

Reply:
I don't really understand when `#[inline]` is necessary. Quoting from https://matklad.github.io/2021/07/09/inline-in-rust.html:

> If a function f is defined in a crate A, then all calls to f from within A can be inlined, as the compiler has full access to f. If, however, f is called from some downstream crate B, such calls can’t be inlined.

> That’s where the main usage of #[inline] comes from — it enables cross-crate inlining.

> Generic functions are implicitly inlinable.

So `init_component` should never need an explicit `#[inline]` since it is already generic, and `init_component_inner`... if a downstream crate compiles `init_component` then rustc has no access to the body of `init_component`, so we should mark it as `#[inline]` if we want component initialization to be fully inlined?

I'll add it.

Reply:
Indeed